### ISSUE TITLE
Allow installation of plugins from behind a proxy

### ISSUE THREAD

**Issue by FergusNelson** [elastic/puppet-elasticsearch#152]
The elasticsearch plugin command does not use environment proxy settings for retrieving the plugins. However it does support command line args for proxy settings.  

/usr/share/elasticsearch/bin/plugin -DproxyPort=$proxy_port -DproxyHost=$proxy_host 

The plugin resource should expose these parameters explicitly. As a work around i have overridden the $plugintool parameter 

$plugintool = "/usr/share/elasticsearch/bin/plugin -DproxyPort=$proxy_port -DproxyHost=$proxy_host "

**Comment by muelli** [comment]
It doesn't seem to work for me.
How is that supposed to work for proxies requiring authentication?  What values should I put for `proxy_host` and `proxy_port`?

**Comment by electrical** [comment]
@muelli i think you could do something like this for the auth part: `proxy_host = 'user:pass@host'`

**Comment by muelli** [comment]
hm. that doesn't seem to work for me :-(
Also, https://github.com/elastic/elasticsearch/commit/cf7ebfcebf50ad596bf5d6f373da94f01500daf9 mentions `-Dhttp.proxyHost`, not `-Dproxy_host`. But it remains to be seem if there is a working parameter for authenticated proxies. So far, I haven't managed to get it working.

Why can the plugin installer not simply take the environment variable? I've filed a bug about it: https://github.com/elastic/elasticsearch/issues/11001

**Comment by joanmarcriera** [comment]
Same problem here .

**Comment by muelli** [comment]
the problem is that the elastic people seem to be unwilling to respect that GNU/Linux hosts are indeed a platform that people run their ES instances on. At least they deem it not important enough to support the proxy use case as can be seen in the other bug report https://github.com/elastic/elasticsearch/issues/16633 which might also contain a workaround.

**Comment by tylerjl** [comment]
What's unclear to me is whether the java property needs the `http.` prepended. According to [our docs](https://www.elastic.co/guide/en/elasticsearch/plugins/2.2/_other_command_line_parameters.html#_proxy_settings), the option is `-DproxyHost` for `bin/plugin`, but the [Java 7](http://docs.oracle.com/javase/7/docs/technotes/guides/net/proxies.html) and [Java 8](https://docs.oracle.com/javase/8/docs/technotes/guides/net/proxies.html) docs only mention `-Dhttp.proxyHost` and I can't find any special handling in the `bin/plugin` code that would translate those into `http.proxyHost`. Maybe the fix here is a simple change to `http.proxyHost`?

Does someone want to confirm whether a `bin/plugin` command using `-Dhttp.proxyHost` instead of `-DproxyHost` works for them? I'd set something up but would feel better about it if it's confirmed working out in the field. **If you need an authenticated proxy**, maybe try [this approach](http://stackoverflow.com/a/1626625) and if that works, we can add some additional parameters to use it.

**Comment by sysgen** [comment]
For reference, it appears to work fine with the protocol prefix (https in this case) and an authenticating proxy:

````
D:\foo>elasticsearch-2.3.4\bin\plugin -Dhttps.proxyHost=myproxy.domain -Dhttps.proxyPort=80 -Dhttps.proxyUser=domain\userid -Dhttps.proxyPassword=mypassword install shield
-> Installing shield...
Trying https://download.elastic.co/elasticsearch/release/org/elasticsearch/plugin/shield/2.3.4/shield-2.3.4.zip ...
Downloading ..............................................................................................................................................DONE
Verifying https://download.elastic.co/elasticsearch/release/org/elasticsearch/plugin/shield/2.3.4/shield-2.3.4.zip checksums if available ...
Downloading .DONE
Installed shield into D:\foo\elasticsearch-2.3.4\plugins\shield
````

**Comment by tylerjl** [comment]
It does seem to appear that the actual Elasticsearch docs have been wrong all along; I've submitted a PR (see above) and if that seems to be the case this should be a simple case of just changing the Java properties that we're passing into the `plugin` command.

**Comment by tylerjl** [comment]
Verified that this is [a documentation error](https://github.com/elastic/elasticsearch/pull/19503), so I'll be making a simple fix in this module to reflect the correct options (for both HTTP and HTTPS).

**Comment by muelli** [comment]
How does it work with a proxy requiring authentication?

**Comment by tylerjl** [comment]
@muelli Ah, I missed that someone had confirmed those Java properties work. I'll implement those parameters as well.

**Comment by tylerjl** [comment]
Implemented in #670, should be in the next release. I refactored how proxy parameters are passed in quite a bit, it should be much more predictable now that we're parsing things via `URL()` rather than a regex.

**Comment by acidwars** [comment]
I changed the elasticsearch-plugin and changed the last line:

```
exec "$JAVA" $ES_JAVA_OPTS -DproxyHost=proxy_host -DproxyPort=port -Delasticsearch "${path_props[@]}" -cp "$ES_HOME/lib/*" org.elasticsearch.plugins.PluginCli "${args[@]}"
~
```

Works for me!
